src/pip/_internal/network/auth.py

```diff
@@ -359,7 +359,7 @@ def _prompt_for_password(
 
     # Factored out to allow for easy patching in tests
     def _should_save_password_to_keyring(self) -> bool:
-        if get_keyring_provider() is None:
+        if isinstance(get_keyring_provider(), KeyRingNullProvider):
             return False
         return ask("Save credentials to keyring [y/N]: ", ["y", "n"]) == "y"
```

Comment:
Maybe we should refactor this whole thing into the provider classes. Checking isinstance here feels like an anti-pattern.

Reply:
Updated, but may need to suggest a better name